### DISCUSSION TITLE
Any way to check for disconnection without side effects (like discarding request body)

### DISCUSSION THREAD

**Discussion by msdrigg** [Kludex/uvicorn#2046] [Q&A]
For some background, I got here looking for a way to cancel costly actions if the client disconnects. I submitted this same discussion to starlette https://github.com/encode/starlette/discussions/2215, but I thought that this might be a better form:

I see that starlette offers an async `is_disconnected` method on their `Request`:

```python
async def is_disconnected(self) -> bool:
    if not self._is_disconnected:
        message: Message = {}

        # If message isn't immediately available, move on
        with anyio.CancelScope() as cs:
            cs.cancel()
            message = await self._receive()

        if message.get("type") == "http.disconnect":
            self._is_disconnected = True

    return self._is_disconnected
```

https://github.com/encode/starlette/blob/844220819f1cfcf2726d1628dd4ea597182d6f3e/starlette/requests.py#L293

Here we see that `is_disconnected` calls `self._receive()` (which is the asgi `receive` function). I believe that this could possibly consume and discard a http.request message containing the request body.

Is there any way to listen for disconnection (ideally create an Event that signals disconnection) without discarding any messages from `receive`? I read the spec, but unless I buffer arbitrarily many messages from `receive`, I don't know how I would do this.

I am relatively new to asgi, so I may be missing something (or not asking in the right place), but I wanted to reach out to see if anyone here could point me in the right direction

**Comment by hari01584** [comment]
Can you see if https://github.com/tiangolo/fastapi/discussions/11360 can work for you?

**Reply by msdrigg** [reply]
So your solution is really good actually. All my attempts just lose the body, while yours keeps it and returns it after buffering. I still see this as limiting though, because it will buffer the entire body, so any kind of streamed request will get buffered entirely in memory and this will remove any kind of tcp window control/backpressure support (https://sumofbytes.com/blog/understanding-tcp-protocol-and-backpressure). 

Most ASGI servers (uvicorn for example: https://www.uvicorn.org/server-behavior/#flow-control) support tcp flow control based off when the client calls the`receive()` method.

**Reply by msdrigg** [reply]
I'd support the PR though and realistically I think it's the best that can be done without an extension to the ASGI protocol with a `.receiveDisconnect()` method or something
